lib/pyfrc/physics/tankmodel.py

```diff
@@ -0,0 +1,263 @@
+        
+        self.acceleration = (appliedVoltage - (self._kv * self.velocity)) / self._ka
+        self.velocity += self.acceleration * tm_diff
+        
+        return self.velocity
+
+
+class TankModel:
+    '''
+        This is a model of a FRC tankdrive-style drivetrain that will provide
+        vaguely realistic motion for the simulator.
+    
+        This drivetrain model makes a number of assumptions:
+        
+        * N motors per side
+        * Constant gearing
+        * Motors are geared together
+        * Wheels do not 'slip' on the ground
+        * Each side of the robot moves in unison
+    
+        There are two ways to construct this model. You can use the theoretical
+        model via :func:`TankModel.theory` and provide robot parameters
+        such as gearing, total mass, etc.
+        
+        Alternatively, if you measure ``kv``, ``ka``, and ``vintercept`` as
+        detailed in the paper mentioned above, you can plug those values in
+        directly instead using the :class:`TankModel` constructor instead. For
+        more information about measuring your own values, see the paper and
+        `this thread on ChiefDelphi <https://www.chiefdelphi.com/forums/showthread.php?t=161539>`_.
+        
+        Example usage for a 90lb robot with 2 CIM motors on each side with 6 inch
+        wheels::
+
+            from pyfrc.physics import motors, tankmodel
+
+            class PhysicsEngine:
+                
+                def __init__(self, physics_controller):
+                    self.physics_controller = physics_controller
+                    
+                    self.drivetrain = tankmodel.TankModel.theory(motors.MOTOR_CFG_CIM,
+                                                                 robot_mass=90, gearing=10.71,
+                                                                 nmotors=2, x_wheelbase=2.0,
+                                                                 wheel_diameter=0.5)
+                    
+                def update_sim(self, hal_data, now, tm_diff):
+                    # TODO: get motor values from hal_data
+                    velocity, rotation = self.drivetrain.get_vector(l_motor, r_motor)
+                    self.physics_controller.drive(velocity, rotation, tm_diff)
+                    
+                    # optional: compute encoder
+                    # l_encoder = self.drivetrain.l_position * ENCODER_TICKS_PER_FT
+                    # r_encoder = self.drivetrain.r_position * ENCODER_TICKS_PER_FT
+    '''
+    
+    @classmethod
+    def theory(cls,
+               motor_config: MotorModelConfig,
+               robot_mass: float,
+               gearing: float,
+               nmotors: int=1,
+               x_wheelbase: float=_kitbot_wheelbase,
+               robot_width: float=_kitbot_width,
+               robot_length: float=_kitbot_length,
+               wheel_diameter: float=0.5,
```

Comment:
Have you guys considered having Travis run a formatter to check for these things? WPILib uses yapf, which supports pep8, but it makes the args with type hints harder to read for this file because it puts several args on the same line. I'm pretty sure other formatters exist though.

Reply:
It keeps getting suggested, but nobody seems to want to do it, and I don't care enough to do it. :)